File: server/assets/js/world_builder/AnthropicClient.js
/**
 * Anthropic Client for World Builder
 *
 * Client-side API calls to Anthropic with streaming support.
 * Uses BYOK (Bring Your Own Key) - key stored in localStorage.
 */

const API_URL = 'https://api.anthropic.com/v1/messages'
const API_VERSION = '2023-06-01'

/**
 * Get the API key from localStorage
 * @returns {string|null} The decrypted API key or null
 */
export function getApiKey() {
  const encrypted = localStorage.getItem('anthropic_api_key_encrypted')
  if (!encrypted) return null

  try {
    return atob(encrypted).replace('loka_wb_', '')
  } catch {
    return null
  }
}

/**
 * Check if API key is configured
 * @returns {boolean}
 */
export function isConfigured() {
  return getApiKey() !== null
}

/**
 * Send a message to Claude with streaming
 *
 * @param {Object} options
 * @param {string} options.model - Model to use (default: claude-opus-4-5-20251101)
 * @param {Array} options.messages - Conversation messages
 * @param {string} options.system - System prompt
 * @param {Array} options.tools - Tool definitions (optional)
 * @param {number} options.maxTokens - Max tokens in response (default: 4096)
 * @param {function} options.onText - Callback for text chunks
 * @param {function} options.onToolUse - Callback for tool use blocks
 * @param {function} options.onComplete - Callback when done
 * @param {function} options.onError - Callback on error
 * @returns {Promise<Object>} Final response
 */
export async function streamMessage({
  model = 'claude-opus-4-5-20251101',
  messages,
  system,
  tools = [],
  maxTokens = 4096,
  onText = () => {},
  onToolUse = () => {},
  onComplete = () => {},
  onError = () => {}
}) {
  const apiKey = getApiKey()
  if (!apiKey) {
    const error = new Error('API key not configured')
    onError(error)
    throw error
  }

  const body = {
    model,
    max_tokens: maxTokens,
    messages,
    stream: true
  }

  if (system) {
    body.system = system
  }

  if (tools && tools.length > 0) {
    body.tools = tools
  }

  try {
    const response = await fetch(API_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': apiKey,
        'anthropic-version': API_VERSION,
        'anthropic-dangerous-direct-browser-access': 'true'
      },
      body: JSON.stringify(body)
    })

    if (!response.ok) {
      const errorBody = await response.text()
      let errorMessage = `API error: ${response.status}`
      try {
        const errorJson = JSON.parse(errorBody)
        errorMessage = errorJson.error?.message || errorMessage
      } catch {}
      const error = new Error(errorMessage)
      onError(error)
      throw error
    }

    // Parse SSE stream
    const reader = response.body.getReader()
    const decoder = new TextDecoder()
    let buffer = ''
    let fullText = ''
    let toolUseBlocks = []
    let currentToolUse = null
    let inputUsage = 0
    let outputUsage = 0
    let stopReason = null

    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })

      // Process complete lines
      const lines = buffer.split('\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (!line.startsWith('data: ')) continue
        const data = line.slice(6)
        if (data === '[DONE]') continue

        try {
          const event = JSON.parse(data)

          switch (event.type) {
            case 'message_start':
              inputUsage = event.message?.usage?.input_tokens || 0
              break

            case 'content_block_start':
              if (event.content_block?.type === 'tool_use') {
                currentToolUse = {
                  id: event.content_block.id,
                  name: event.content_block.name,
                  input: ''
                }
              }
              break

            case 'content_block_delta':
              if (event.delta?.type === 'text_delta') {
                const text = event.delta.text || ''
                fullText += text
                onText(text, fullText)
              } else if (event.delta?.type === 'input_json_delta') {
                if (currentToolUse) {
                  currentToolUse.input += event.delta.partial_json || ''
                }
              }
              break

            case 'content_block_stop':
              if (currentToolUse) {
                // Parse the accumulated JSON input
                try {
                  currentToolUse.input = JSON.parse(currentToolUse.input)
                } catch {
                  currentToolUse.input = {}
                }
                toolUseBlocks.push(currentToolUse)
                onToolUse(currentToolUse)
                currentToolUse = null
              }
              break

            case 'message_delta':
              stopReason = event.delta?.stop_reason
              outputUsage = event.usage?.output_tokens || 0
              break

            case 'message_stop':
              // Message complete
              break

            case 'error':
              const error = new Error(event.error?.message || 'Unknown error')
              onError(error)
              throw error
          }
        } catch (parseError) {
          // Ignore parse errors for malformed events
          if (parseError.message !== 'Unknown error') {
            console.warn('Failed to parse SSE event:', parseError)
          }
        }
      }
    }

    const result = {
      text: fullText,
      toolUse: toolUseBlocks,
      stopReason,
      usage: {
        inputTokens: inputUsage,
        outputTokens: outputUsage,
        totalTokens: inputUsage + outputUsage
      }
    }

    onComplete(result)
    return result

  } catch (error) {
    onError(error)
    throw error
  }
}

/**
 * Send a message without streaming (simpler API)
 *
 * @param {Object} options - Same as streamMessage but without streaming callbacks
 * @returns {Promise<Object>} Response with text and tool use
 */
export async function sendMessage(options) {
  return streamMessage({
    ...options,
    onText: () => {},
    onToolUse: () => {},
    onComplete: () => {},
    onError: () => {}
  })
}

/**
 * Estimate token count for a message (rough approximation)
 * Uses ~4 characters per token as rough estimate
 *
 * @param {string} text - Text to estimate
 * @returns {number} Estimated token count
 */
export function estimateTokens(text) {
  if (!text) return 0
  return Math.ceil(text.length / 4)
}

/**
 * Format messages for the API
 *
 * @param {Array} messages - Array of {role, content} or {role, content: [{type, ...}]}
 * @returns {Array} Formatted messages
 */
export function formatMessages(messages) {
  return messages.map(msg => {
    if (typeof msg.content === 'string') {
      return { role: msg.role, content: msg.content }
    }
    return msg
  })
}

export default {
  getApiKey,
  isConfigured,
  streamMessage,
  sendMessage,
  estimateTokens,
  formatMessages
}

File: server/lib/loka/world_builder/anthropic_client.ex
defmodule Loka.WorldBuilder.AnthropicClient do
  @moduledoc """
  Server-side Anthropic API client for the World Builder.

  Provides streaming chat completions with tool use support.
  Used by the LiveView chat component.
  """

  require Logger

  @api_url "https://api.anthropic.com/v1/messages"
  @api_version "2023-06-01"

  # Configurable via application config:
  # config :loka, Loka.WorldBuilder.AnthropicClient,
  #   default_model: "claude-sonnet-4-20250514",
  #   default_max_tokens: 4096,
  #   timeout: 120_000

  defp config(key, default) do
    Application.get_env(:loka, __MODULE__, [])
    |> Keyword.get(key, default)
  end

  defp default_model, do: config(:default_model, "claude-sonnet-4-20250514")
  defp default_max_tokens, do: config(:default_max_tokens, 4096)
  defp default_timeout, do: config(:timeout, 120_000)

  @doc """
  Send a message to Claude and stream the response.

  ## Parameters
    - messages: List of message maps with role and content
    - tools: List of tool definitions (optional)
    - opts: Options including:
      - :model - Model to use (default: claude-sonnet-4-20250514)
      - :max_tokens - Max tokens (default: 4096)
      - :system - System prompt
      - :on_text - Callback for text deltas: fn text -> ... end
      - :on_tool_use - Callback for tool calls: fn tool_name, tool_input -> ... end
      - :on_done - Callback when complete: fn response -> ... end
      - :on_error - Callback for errors: fn error -> ... end

  ## Returns
    - {:ok, response} on success
    - {:error, reason} on failure
  """
  def chat(messages, tools \\ [], opts \\ []) do
    api_key = get_api_key()

    if is_nil(api_key) do
      {:error, "ANTHROPIC_API_KEY not configured"}
    else
      do_chat(messages, tools, opts, api_key)
    end
  end

  @doc """
  Send a message and stream responses to a LiveView process.

  ## Parameters
    - lv_pid: LiveView process to send events to
    - messages: List of message maps
    - tools: List of tool definitions
    - opts: Additional options

  Events sent to LiveView:
    - {:anthropic_text_delta, text}
    - {:anthropic_tool_use, tool_name, tool_id, input}
    - {:anthropic_done, response}
    - {:anthropic_error, reason}
  """
  def stream_to_liveview(lv_pid, messages, tools \\ [], opts \\ []) do
    opts =
      opts
      |> Keyword.put(:on_text, fn text ->
        send(lv_pid, {:anthropic_text_delta, text})
      end)
      |> Keyword.put(:on_tool_use, fn tool_name, tool_id, input ->
        send(lv_pid, {:anthropic_tool_use, tool_name, tool_id, input})
      end)
      |> Keyword.put(:on_done, fn response ->
        send(lv_pid, {:anthropic_done, response})
      end)
      |> Keyword.put(:on_error, fn error ->
        send(lv_pid, {:anthropic_error, error})
      end)

    # Run in a separate process to not block LiveView
    Task.start(fn ->
      chat(messages, tools, opts)
    end)
  end

  defp do_chat(messages, tools, opts, api_key) do
    model = Keyword.get(opts, :model, default_model())
    max_tokens = Keyword.get(opts, :max_tokens, default_max_tokens())
    system = Keyword.get(opts, :system)

    body =
      %{
        model: model,
        max_tokens: max_tokens,
        messages: format_messages(messages),
        stream: true
      }
      |> maybe_add_system(system)
      |> maybe_add_tools(tools)

    headers = [
      {"Content-Type", "application/json"},
      {"x-api-key", api_key},
      {"anthropic-version", @api_version}
    ]

    Logger.info("[AnthropicClient] Starting chat with model: #{model}")

    case stream_request(@api_url, headers, body, opts) do
      {:ok, response} ->
        Logger.info("[AnthropicClient] Chat completed successfully")
        {:ok, response}

      {:error, reason} = error ->
        Logger.error("[AnthropicClient] Chat failed: #{inspect(reason)}")
        if on_error = opts[:on_error], do: on_error.(reason)
        error
    end
  end

  defp format_messages(messages) do
    Enum.map(messages, fn msg ->
      %{
        role: msg[:role] || msg["role"],
        content: msg[:content] || msg["content"]
      }
    end)
  end

  defp maybe_add_system(body, nil), do: body
  defp maybe_add_system(body, system), do: Map.put(body, :system, system)

  defp maybe_add_tools(body, []), do: body

  defp maybe_add_tools(body, tools) do
    formatted_tools =
      Enum.map(tools, fn tool ->
        %{
          name: tool[:name] || tool["name"],
          description: tool[:description] || tool["description"],
          input_schema: tool[:input_schema] || tool[:inputSchema] || tool["input_schema"]
        }
      end)

    Map.put(body, :tools, formatted_tools)
  end

  defp stream_request(url, headers, body, opts) do
    # Use Req for streaming HTTP with callback function
    on_text = opts[:on_text]
    on_tool_use = opts[:on_tool_use]
    on_done = opts[:on_done]

    # Accumulate response state
    initial_state = %{
      text: "",
      tool_uses: [],
      current_tool: nil,
      stop_reason: nil,
      buffer: ""
    }

    # Use into: with a callback function for proper Req 0.5.x streaming
    stream_fn = fn {:data, data}, {req, resp} ->
      # Process SSE data and update state
      state = Process.get(:anthropic_stream_state, initial_state)
      new_state = process_sse_data(data, state, on_text, on_tool_use)
      Process.put(:anthropic_stream_state, new_state)
      {:cont, {req, resp}}
    end

    # Initialize state in process dictionary
    Process.put(:anthropic_stream_state, initial_state)

    request =
      Req.new(
        url: url,
        method: :post,
        headers: headers,
        json: body,
        receive_timeout: default_timeout(),
        into: stream_fn
      )

    case Req.request(request) do
      {:ok, %Req.Response{status: 200}} ->
        # Get final state and clean up
        final_state = Process.get(:anthropic_stream_state, initial_state)
        Process.delete(:anthropic_stream_state)

        response = %{
          text: final_state.text,
          tool_uses: final_state.tool_uses,
          stop_reason: final_state.stop_reason
        }

        if on_done, do: on_done.(response)
        {:ok, response}

      {:ok, %Req.Response{status: status, body: body}} ->
        Process.delete(:anthropic_stream_state)
        {:error, "API error: #{status} - #{inspect(body)}"}

      {:error, reason} ->
        Process.delete(:anthropic_stream_state)
        {:error, "Request failed: #{inspect(reason)}"}
    end
  end

  defp process_sse_data(data, state, on_text, on_tool_use) do
    data
    |> String.split("\n")
    |> Enum.reduce(state, fn line, acc ->
      process_sse_line(line, acc, on_text, on_tool_use)
    end)
  end

  defp process_sse_line("data: " <> json_data, state, on_text, on_tool_use) do
    case Jason.decode(json_data) do
      {:ok, event} ->
        process_event(event, state, on_text, on_tool_use)

      {:error, _} ->
        state
    end
  end

  defp process_sse_line(_, state, _, _), do: state

  defp process_event(%{"type" => "content_block_start", "content_block" => block}, state, _, _) do
    case block do
      %{"type" => "tool_use", "id" => id, "name" => name} ->
        %{state | current_tool: %{id: id, name: name, input: ""}}

      _ ->
        state
    end
  end

  defp process_event(%{"type" => "content_block_delta", "delta" => delta}, state, on_text, _) do
    case delta do
      %{"type" => "text_delta", "text" => text} ->
        if on_text, do: on_text.(text)
        %{state | text: state.text <> text}

      %{"type" => "input_json_delta", "partial_json" => json} ->
        if state.current_tool do
          current = state.current_tool
          %{state | current_tool: %{current | input: current.input <> json}}
        else
          state
        end

      _ ->
        state
    end
  end

  defp process_event(%{"type" => "content_block_stop"}, state, _, on_tool_use) do
    if state.current_tool do
      tool = state.current_tool

      input =
        case Jason.decode(tool.input) do
          {:ok, parsed} -> parsed
          {:error, _} -> %{}
        end

      if on_tool_use, do: on_tool_use.(tool.name, tool.id, input)

      tool_use = %{
        id: tool.id,
        name: tool.name,
        input: input
      }

      %{state | tool_uses: state.tool_uses ++ [tool_use], current_tool: nil}
    else
      state
    end
  end

  defp process_event(
         %{"type" => "message_delta", "delta" => %{"stop_reason" => reason}},
         state,
         _,
         _
       ) do
    %{state | stop_reason: reason}
  end

  defp process_event(_, state, _, _), do: state

  defp get_api_key do
    Application.get_env(:loka, :anthropic_api_key) ||
      System.get_env("ANTHROPIC_API_KEY")
  end

  @doc """
  Get tool definitions in Anthropic API format.
  """
  def get_tools do
    alias Loka.WorldBuilder.MCP.Tools

    Tools.tools()
    |> Enum.map(fn tool ->
      %{
        name: tool.name,
        description: String.trim(tool.description),
        input_schema: tool.inputSchema
      }
    end)
  end
end
